File: fleet-integration/fleet_integration/device.py
from __future__ import annotations

import json
import logging
import random
import threading
import time
from collections.abc import Callable
from typing import Any, Optional

import paho.mqtt.client as mqtt

from fleet_integration.config import DeviceConfig
from fleet_integration.models import CommandPayload, CommandResponse, DeviceStats

logger = logging.getLogger(__name__)

# Type aliases
StatsGenerator = Callable[[], DeviceStats]
CommandHandler = Callable[[CommandPayload], dict[str, Any]]


class FleetDevice:
    """MQTT-based fleet device SDK.

    Handles:
    - Connecting (with automatic retry) to the MQTT broker.
    - Periodically publishing telemetry stats.
    - Receiving commands and dispatching them to registered handlers.
    - Publishing command responses.

    Basic usage
    -----------
    >>> device = FleetDevice()
    >>> device.run()          # blocks forever, publishes random stats

    Custom stats
    ------------
    >>> @device.stats_generator
    ... def my_stats() -> DeviceStats:
    ...     return DeviceStats(
    ...         device_id=device.device_id,
    ...         sensor1=read_temp(),
    ...         sensor2=read_humidity(),
    ...         sensor3=read_pressure(),
    ...         version=device.version,
    ...     )

    Custom command handler
    ----------------------
    >>> @device.command("ping")
    ... def handle_ping(cmd: CommandPayload) -> dict:
    ...     return {"message": "pong"}

    Async-friendly
    --------------
    Call `device.start()` + `device.stop()` instead of `device.run()` when you
    need the loop to run in a background thread and keep control of the main
    thread yourself.
    """

    def __init__(self, config: Optional[DeviceConfig] = None) -> None:
        self._config = config or DeviceConfig()
        self._version = self._config.version

        self._command_handlers: dict[str, CommandHandler] = {}
        self._stats_gen: StatsGenerator = self._default_stats

        self._client: Optional[mqtt.Client] = None
        self._stop_event = threading.Event()
        self._stats_thread: Optional[threading.Thread] = None

        # Register built-in command handlers
        self._register_builtin_handlers()

    # ------------------------------------------------------------------
    # Public properties
    # ------------------------------------------------------------------

    @property
    def device_id(self) -> str:
        return self._config.device_id

    @property
    def version(self) -> str:
        return self._version

    # ------------------------------------------------------------------
    # Decorator API
    # ------------------------------------------------------------------

    def command(self, name: str) -> Callable[[CommandHandler], CommandHandler]:
        """Register a handler for the given command name.

        The decorated function receives a :class:`CommandPayload` and must
        return a plain ``dict`` that will be set as the ``response`` field of
        the published :class:`CommandResponse`.

        >>> @device.command("ping")
        ... def ping(cmd: CommandPayload) -> dict:
        ...     return {"message": "pong"}
        """

        def decorator(fn: CommandHandler) -> CommandHandler:
            self._command_handlers[name] = fn
            return fn

        return decorator

    def stats_generator(self, fn: StatsGenerator) -> StatsGenerator:
        """Replace the default random stats generator with a custom one.

        The decorated function must return a :class:`DeviceStats` instance.

        >>> @device.stats_generator
        ... def my_stats() -> DeviceStats:
        ...     return DeviceStats(device_id=device.device_id, ...)
        """
        self._stats_gen = fn
        return fn

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def run(self) -> None:
        """Connect, start the MQTT loop and block until KeyboardInterrupt."""
        self._connect()
        self._client.loop_start()  # type: ignore[union-attr]
        self._stats_thread = threading.Thread(target=self._publish_stats_loop, daemon=True)
        self._stats_thread.start()
        logger.info(f"[{self.device_id}] Running. Press Ctrl-C to stop.")
        try:
            while not self._stop_event.is_set():
                time.sleep(0.5)
        except KeyboardInterrupt:
            pass
        finally:
            self.stop()

    def start(self) -> None:
        """Connect and start background threads without blocking."""
        self._connect()
        self._client.loop_start()  # type: ignore[union-attr]
        self._stats_thread = threading.Thread(target=self._publish_stats_loop, daemon=True)
        self._stats_thread.start()

    def stop(self) -> None:
        """Disconnect and stop all background threads."""
        self._stop_event.set()
        if self._client is not None:
            self._client.loop_stop()
            self._client.disconnect()
        logger.info(f"[{self.device_id}] Stopped.")

    # ------------------------------------------------------------------
    # Internal – MQTT wiring
    # ------------------------------------------------------------------

    def _connect(self) -> None:
        self._client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2)
        self._client.on_connect = self._on_connect
        self._client.on_message = self._on_message

        while not self._stop_event.is_set():
            try:
                self._client.connect(self._config.broker, self._config.port, 60)
                break
            except Exception as exc:
                logger.warning(
                    f"[{self.device_id}] Cannot connect to broker: {exc}. "
                    f"Retrying in {self._config.reconnect_delay}s…"
                )
                time.sleep(self._config.reconnect_delay)

    def _on_connect(
        self,
        client: mqtt.Client,
        userdata: Any,
        flags: Any,
        rc: int,
        properties: Any = None,
    ) -> None:
        logger.info(f"[{self.device_id}] Connected to broker (rc={rc})")
        client.subscribe(f"fleet/{self.device_id}/commands")

    def _on_message(self, client: mqtt.Client, userdata: Any, msg: mqtt.MQTTMessage) -> None:
        try:
            raw = json.loads(msg.payload.decode())
            cmd = CommandPayload(**raw)
            logger.info(f"[{self.device_id}] Received command: {cmd.command} (id={cmd.command_id})")
            response_data = self._dispatch(cmd)
            response = CommandResponse(
                device_id=self.device_id,
                command_id=cmd.command_id,
                response=response_data,
            )
            client.publish(
                f"fleet/{self.device_id}/commands/response",
                response.model_dump_json(),
            )
        except Exception as exc:
            logger.error(f"[{self.device_id}] Error handling command: {exc}")

    # ------------------------------------------------------------------
    # Internal – command dispatch
    # ------------------------------------------------------------------

    def _dispatch(self, cmd: CommandPayload) -> dict[str, Any]:
        handler = self._command_handlers.get(cmd.command)
        if handler is None:
            return {"message": f"Command '{cmd.command}' acknowledged"}
        return handler(cmd)

    def _register_builtin_handlers(self) -> None:
        @self.command("ping")
        def _ping(cmd: CommandPayload) -> dict:
            return {"message": "pong"}

        @self.command("reboot")
        def _reboot(cmd: CommandPayload) -> dict:
            return {"message": "Rebooting device", "eta_seconds": random.randint(5, 30)}

        @self.command("reset_sensors")
        def _reset_sensors(cmd: CommandPayload) -> dict:
            return {"message": "Sensors reset", "values": [0.0, 0.0, 0.0]}

        @self.command("report_full")
        def _report_full(cmd: CommandPayload) -> dict:
            return {
                "message": "Full report",
                "sensor1": round(random.uniform(0, 100), 2),
                "sensor2": round(random.uniform(20, 80), 2),
                "sensor3": round(random.uniform(-10, 50), 2),
                "uptime_seconds": random.randint(100, 100000),
            }

        @self.command("update_software")
        def _update_software(cmd: CommandPayload) -> dict:
            target = cmd.payload.get("version")
            new_version = target if target else _bump_version(self._version)
            self._version = new_version
            return {"message": "Software updated", "version": new_version}

    # ------------------------------------------------------------------
    # Internal – stats publishing
    # ------------------------------------------------------------------

    def _publish_stats_loop(self) -> None:
        logger.info(f"[{self.device_id}] Starting stats publish loop…")
        while not self._stop_event.is_set():
            try:
                stats = self._stats_gen()
                topic = f"fleet/{self.device_id}/stats"
                if self._client is not None:
                    self._client.publish(topic, stats.model_dump_json())
                    logger.debug(f"[{self.device_id}] Published stats: {stats}")
            except Exception as exc:
                logger.error(f"[{self.device_id}] Error publishing stats: {exc}")

            interval = random.uniform(
                self._config.stats_interval_min,
                self._config.stats_interval_max,
            )
            self._stop_event.wait(interval)

    def _default_stats(self) -> DeviceStats:
        return DeviceStats(
            device_id=self.device_id,
            sensor1=round(random.uniform(0, 100), 2),
            sensor2=round(random.uniform(20, 80), 2),
            sensor3=round(random.uniform(-10, 50), 2),
            version=self._version,
        )


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _bump_version(version: str) -> str:
    """Increment the patch segment of a semver string (e.g. v1.0.1 → v1.0.2)."""
    try:
        prefix = version[0] if version[0].isalpha() else ""
        parts = version.lstrip("vV").split(".")
        parts[-1] = str(int(parts[-1]) + 1)
        return prefix + ".".join(parts)
    except Exception:
        return version

File: fleet-integration-rs/src/device.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex, RwLock};
use std::time::Duration;

use rand::Rng;
use rumqttc::{AsyncClient, Event, MqttOptions, Packet, QoS};
use serde_json::{json, Value};
use tracing::{debug, error, info, warn};

use crate::config::DeviceConfig;
use crate::models::{CommandPayload, CommandResponse, DeviceStats};

// ---------------------------------------------------------------------------
// Type aliases
// ---------------------------------------------------------------------------

type CommandHandler = Arc<dyn Fn(&CommandPayload) -> Value + Send + Sync>;
type StatsGenerator = Arc<dyn Fn() -> DeviceStats + Send + Sync>;

// ---------------------------------------------------------------------------
// FleetDevice
// ---------------------------------------------------------------------------

/// MQTT-based fleet device SDK.
///
/// Handles:
/// - Connecting (with automatic retry) to the MQTT broker.
/// - Periodically publishing telemetry stats.
/// - Receiving commands and dispatching them to registered handlers.
/// - Publishing command responses.
///
/// # Basic usage
///
/// ```no_run
/// use fleet_integration::{FleetDevice, DeviceConfig};
///
/// #[tokio::main]
/// async fn main() {
///     let device = FleetDevice::new(DeviceConfig::default());
///     device.run().await.unwrap();
/// }
/// ```
///
/// # Custom stats
///
/// ```no_run
/// # use fleet_integration::{FleetDevice, DeviceConfig, DeviceStats};
/// # let device = FleetDevice::new(DeviceConfig::default());
/// device.set_stats_generator({
///     let id = device.device_id().to_string();
///     let ver = device.version();
///     move || DeviceStats::now(&id, 42.0, 65.0, 1013.0, Some(ver.clone()))
/// });
/// ```
///
/// # Custom command handler
///
/// ```no_run
/// # use fleet_integration::FleetDevice;
/// # use serde_json::json;
/// # let device = FleetDevice::new(Default::default());
/// device.on_command("ping", |_cmd| json!({"message": "pong"}));
/// ```
///
/// # Background usage
///
/// ```no_run
/// # use fleet_integration::FleetDevice;
/// # use std::sync::Arc;
/// let device = Arc::new(FleetDevice::new(Default::default()));
/// let handle = {
///     let d = Arc::clone(&device);
///     tokio::spawn(async move { d.run().await })
/// };
/// // ... do other work ...
/// handle.abort();
/// ```
pub struct FleetDevice {
    config: DeviceConfig,
    version: Arc<Mutex<String>>,
    handlers: Arc<RwLock<HashMap<String, CommandHandler>>>,
    stats_gen: Arc<RwLock<StatsGenerator>>,
}

impl FleetDevice {
    /// Create a new device with the given configuration.
    /// Built-in command handlers (`ping`, `reboot`, `reset_sensors`,
    /// `report_full`, `update_software`) are registered automatically.
    pub fn new(config: DeviceConfig) -> Self {
        let version = Arc::new(Mutex::new(config.version.clone()));
        let handlers = Arc::new(RwLock::new(HashMap::<String, CommandHandler>::new()));

        let device = Self {
            stats_gen: Arc::new(RwLock::new(Self::make_default_stats_gen(
                &config.device_id,
                Arc::clone(&version),
            ))),
            config,
            version,
            handlers,
        };

        device.register_builtin_handlers();
        device
    }

    // ------------------------------------------------------------------
    // Public API
    // ------------------------------------------------------------------

    pub fn device_id(&self) -> &str {
        &self.config.device_id
    }

    pub fn version(&self) -> String {
        self.version.lock().unwrap().clone()
    }

    /// Register (or replace) a handler for the given command name.
    ///
    /// The closure receives a [`CommandPayload`] and must return a
    /// [`serde_json::Value`] that will be set as the `response` field
    /// of the published [`CommandResponse`].
    pub fn on_command<F>(&self, name: impl Into<String>, handler: F)
    where
        F: Fn(&CommandPayload) -> Value + Send + Sync + 'static,
    {
        self.handlers
            .write()
            .unwrap()
            .insert(name.into(), Arc::new(handler));
    }

    /// Replace the default random stats generator with a custom one.
    ///
    /// The closure must return a [`DeviceStats`] instance and will be
    /// called on every stats publish interval.
    pub fn set_stats_generator<F>(&self, generator: F)
    where
        F: Fn() -> DeviceStats + Send + Sync + 'static,
    {
        *self.stats_gen.write().unwrap() = Arc::new(generator);
    }

    // ------------------------------------------------------------------
    // Lifecycle
    // ------------------------------------------------------------------

    /// Connect to the broker and run forever, blocking the current task.
    /// Handles Ctrl-C gracefully.
    pub async fn run(&self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let (client, mut eventloop) = self.create_mqtt_client();
        let client = Arc::new(client);

        // Spawn the periodic stats publisher
        {
            let stats_client = Arc::clone(&client);
            let stats_gen = Arc::clone(&self.stats_gen);
            let device_id = self.config.device_id.clone();
            let interval_min = self.config.stats_interval_min;
            let interval_max = self.config.stats_interval_max;

            tokio::spawn(async move {
                info!("[{device_id}] Starting stats publish loop…");
                loop {
                    let stats = { stats_gen.read().unwrap().clone()() };
                    match serde_json::to_string(&stats) {
                        Ok(payload) => {
                            let topic = format!("fleet/{device_id}/stats");
                            if let Err(e) = stats_client
                                .publish(&topic, QoS::AtLeastOnce, false, payload)
                                .await
                            {
                                error!("[{device_id}] Failed to publish stats: {e}");
                            } else {
                                debug!("[{device_id}] Published stats");
                            }
                        }
                        Err(e) => error!("[{device_id}] Failed to serialize stats: {e}"),
                    }

                    let interval = rand::thread_rng().gen_range(interval_min..interval_max);
                    tokio::time::sleep(Duration::from_secs_f64(interval)).await;
                }
            });
        }

        // Drive the MQTT event loop
        loop {
            match eventloop.poll().await {
                Ok(Event::Incoming(Packet::ConnAck(_))) => {
                    info!("[{}] Connected to broker", self.config.device_id);
                    let topic = format!("fleet/{}/commands", self.config.device_id);
                    client.subscribe(&topic, QoS::AtLeastOnce).await?;
                }
                Ok(Event::Incoming(Packet::Publish(msg))) => {
                    let client = Arc::clone(&client);
                    let handlers = Arc::clone(&self.handlers);
                    let device_id = self.config.device_id.clone();
                    tokio::spawn(async move {
                        handle_incoming(client, handlers, device_id, msg).await;
                    });
                }
                Ok(_) => {}
                Err(e) => {
                    warn!(
                        "[{}] MQTT error: {e} — retrying in {}s…",
                        self.config.device_id, self.config.reconnect_delay_secs
                    );
                    tokio::time::sleep(Duration::from_secs_f64(
                        self.config.reconnect_delay_secs,
                    ))
                    .await;
                }
            }
        }
    }

    // ------------------------------------------------------------------
    // Internal helpers
    // ------------------------------------------------------------------

    fn create_mqtt_client(&self) -> (AsyncClient, rumqttc::EventLoop) {
        let mut options = MqttOptions::new(
            &self.config.device_id,
            &self.config.broker,
            self.config.port,
        );
        options.set_keep_alive(Duration::from_secs(60));
        AsyncClient::new(options, 10)
    }

    fn make_default_stats_gen(
        device_id: &str,
        version: Arc<Mutex<String>>,
    ) -> StatsGenerator {
        let device_id = device_id.to_string();
        Arc::new(move || {
            let mut rng = rand::thread_rng();
            let round2 = |v: f64| (v * 100.0).round() / 100.0;
            DeviceStats::now(
                &device_id,
                round2(rng.gen_range(0.0_f64..100.0)),
                round2(rng.gen_range(20.0_f64..80.0)),
                round2(rng.gen_range(-10.0_f64..50.0)),
                Some(version.lock().unwrap().clone()),
            )
        })
    }

    fn register_builtin_handlers(&self) {
        self.on_command("ping", |_cmd| json!({"message": "pong"}));

        self.on_command("reboot", |_cmd| {
            let eta = rand::thread_rng().gen_range(5u64..30);
            json!({"message": "Rebooting device", "eta_seconds": eta})
        });

        self.on_command("reset_sensors", |_cmd| {
            json!({"message": "Sensors reset", "values": [0.0, 0.0, 0.0]})
        });

        self.on_command("report_full", |_cmd| {
            let mut rng = rand::thread_rng();
            let round2 = |v: f64| (v * 100.0).round() / 100.0;
            json!({
                "message": "Full report",
                "sensor1": round2(rng.gen_range(0.0_f64..100.0)),
                "sensor2": round2(rng.gen_range(20.0_f64..80.0)),
                "sensor3": round2(rng.gen_range(-10.0_f64..50.0)),
                "uptime_seconds": rng.gen_range(100u64..100_000),
            })
        });

        let version = Arc::clone(&self.version);
        self.on_command("update_software", move |cmd| {
            let target = cmd.payload.get("version").and_then(Value::as_str);
            let new_version = if let Some(v) = target {
                v.to_string()
            } else {
                let current = version.lock().unwrap();
                bump_version(&current)
            };
            *version.lock().unwrap() = new_version.clone();
            json!({"message": "Software updated", "version": new_version})
        });
    }
}

// ---------------------------------------------------------------------------
// Free helpers
// ---------------------------------------------------------------------------

async fn handle_incoming(
    client: Arc<AsyncClient>,
    handlers: Arc<RwLock<HashMap<String, CommandHandler>>>,
    device_id: String,
    msg: rumqttc::Publish,
) {
    let raw = match std::str::from_utf8(&msg.payload) {
        Ok(s) => s,
        Err(e) => {
            error!("[{device_id}] Non-UTF8 payload: {e}");
            return;
        }
    };

    let cmd: CommandPayload = match serde_json::from_str(raw) {
        Ok(c) => c,
        Err(e) => {
            error!("[{device_id}] Failed to parse command: {e}");
            return;
        }
    };

    info!("[{device_id}] Received command: {} (id={})", cmd.command, cmd.command_id);

    let response_data = {
        let map = handlers.read().unwrap();
        match map.get(&cmd.command) {
            Some(handler) => handler(&cmd),
            None => json!({"message": format!("Command '{}' acknowledged", cmd.command)}),
        }
    };

    let response = CommandResponse::executed(&device_id, &cmd.command_id, response_data);

    match serde_json::to_string(&response) {
        Ok(payload) => {
            let topic = format!("fleet/{device_id}/commands/response");
            if let Err(e) = client.publish(&topic, QoS::AtLeastOnce, false, payload).await {
                error!("[{device_id}] Failed to publish response: {e}");
            }
        }
        Err(e) => error!("[{device_id}] Failed to serialize response: {e}"),
    }
}

fn bump_version(version: &str) -> String {
    let prefix = version.chars().next().filter(|c| c.is_alphabetic());
    let stripped = version.trim_start_matches(|c: char| c.is_alphabetic());
    let parts: Vec<&str> = stripped.split('.').collect();
    let bumped_parts: Vec<String> = parts
        .iter()
        .enumerate()
        .map(|(i, part)| {
            if i == parts.len() - 1 {
                part.parse::<u64>()
                    .map(|n| (n + 1).to_string())
                    .unwrap_or_else(|_| part.to_string())
            } else {
                part.to_string()
            }
        })
        .collect();
    let joined = bumped_parts.join(".");
    match prefix {
        Some(p) => format!("{p}{joined}"),
        None => joined,
    }
}
